cluster-autoscaler/cloudprovider/oci/oke/oci/oci_manager.go

```diff
@@ -0,0 +1,477 @@
+					strings.Contains(*node.NodeError.Message, "quota")) {
+				errorClass = cloudprovider.OutOfResourcesErrorClass
+			}
+
+			instances = append(instances, cloudprovider.Instance{
+				Id: *node.Id,
+				Status: &cloudprovider.InstanceStatus{
+					ErrorInfo: &cloudprovider.InstanceErrorInfo{
+						ErrorClass:   errorClass,
+						ErrorCode:    *node.NodeError.Code,
+						ErrorMessage: *node.NodeError.Message,
+					},
+				},
+			})
+
+			continue
+		}
+
+		switch node.LifecycleState {
+		case oke.NodeLifecycleStateDeleted:
+			klog.V(4).Infof("skipping instance is in deleted state: %q", *node.Id)
+		case oke.NodeLifecycleStateDeleting:
+			instances = append(instances, cloudprovider.Instance{
+				Id: *node.Id,
+				Status: &cloudprovider.InstanceStatus{
+					State: cloudprovider.InstanceDeleting,
+				},
+			})
+		case oke.NodeLifecycleStateCreating, oke.NodeLifecycleStateUpdating:
+			instances = append(instances, cloudprovider.Instance{
+				Id: *node.Id,
+				Status: &cloudprovider.InstanceStatus{
+					State: cloudprovider.InstanceCreating,
+				},
+			})
+		case oke.NodeLifecycleStateActive:
+			instances = append(instances, cloudprovider.Instance{
+				Id: *node.Id,
+				Status: &cloudprovider.InstanceStatus{
+					State: cloudprovider.InstanceRunning,
+				},
+			})
+		default:
+			klog.Warningf("instance found in unhandled state: (%q = %v)", *node.Id, node.LifecycleState)
+		}
+	}
+
+	return instances, nil
+}
+
+// GetNodePoolForInstance returns NodePool to which the given instance belongs.
+func (m *ociManagerImpl) GetNodePoolForInstance(instance OciRef) (NodePool, error) {
+	if instance.NodePoolID == "" {
+		klog.V(4).Infof("node pool id missing from reference: %+v", instance)
+
+		// we're looking up an unregistered node, so we can't use node pool id.
+		nodePool, err := m.nodePoolCache.getByInstance(instance.InstanceID)
+		if err != nil {
+			return nil, err
+		}
+
+		return m.staticNodePools[*nodePool.Id], nil
+	}
+
+	np, found := m.staticNodePools[instance.NodePoolID]
+	if !found {
+		klog.Infof("did not find node pool for reference: %+v", instance)
+		return nil, ErrInstanceNodePoolNotFound
+	}
+
+	return np, nil
+}
+
+// GetNodePoolTemplateNode returns a template node for NodePool.
+func (m *ociManagerImpl) GetNodePoolTemplateNode(np NodePool) (*apiv1.Node, error) {
+
+	nodePool, err := m.nodePoolCache.get(np.Id())
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := m.buildNodeFromTemplate(nodePool)
+	if err != nil {
+		return nil, err
+	}
+
+	return node, nil
+}
+
+// GetNodePoolSize gets NodePool size.
+func (m *ociManagerImpl) GetNodePoolSize(np NodePool) (int, error) {
+	return m.nodePoolCache.getSize(np.Id())
+}
+
+// SetNodePoolSize sets NodePool size.
+func (m *ociManagerImpl) SetNodePoolSize(np NodePool, size int) error {
+
+	err := m.nodePoolCache.setSize(np.Id(), size)
+	if err != nil {
+		return err
+	}
+
+	// We do not wait for the work request to finish or nodes become active on purpose. This allows
```

Comment:
However, we _do_ wait in the instancepool implementation - at least for the instances to be provisioned. 

We also check for capacity and quota errors so that the autoscaler to try to scale a different pool (rather than waiting for `max-node-provision-time` and then trying to scale the same pool again).

We may decide that certain functional behaviors are acceptable between the two implementations. However, they should probably be considered on a case by case basis.

Reply:
> However, we do wait in the instancepool implementation - at least for the instances to be provisioned.

I'll set up a meeting with relevant people to discuss how we want to handle this.

> We also check for capacity and quota errors so that the autoscaler to try to scale a different pool (rather than waiting for max-node-provision-time and then trying to scale the same pool again). We added this stuff based on customer feedback.

I do not know if Node Pools provide an easy way to know whether it is failing based on these criteria. But these are a good ideas to add as a future enhancement.

Reply:
I had a discussion with our team regarding adding a wait to the node pool implementation. We have decided not to add a wait to the node pool implementation as the current implementation does not do this, and we did not want to introduce any kind of potential behavior change.

In particular, adding a wait for a scale up that is quickly followed by a scale down should not cause any issues because the cluster autoscaler already has logic that puts the scale down into a backoff for X minutes after a scale up is performed.

However, in the case where a scale up is fast followed by another scale up, introducing a wait may mean that the autoscaler will have to wait for the previous scale to complete before it can make an additional decision to scale up again.